File: java/39.combination-sum.java
/*
 * @lc app=leetcode id=39 lang=java
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList();
        Deque<Integer> q = new ArrayDeque();

        build(candidates, target, res, q, 0, 0);

        return res;
    }

    public void build(int[] candidates, int target, List<List<Integer>> res, Deque<Integer> q, int i, int sum) {
        if(sum == target) {
            res.add(new ArrayList(q));
            return;
        }

        if(sum > target || i >= candidates.length) {
            return;
        }

        build(candidates, target, res, q, i+1, sum);

        q.addLast(candidates[i]);
        build(candidates, target, res, q, i, sum + candidates[i]);
        q.removeLast();
    }
}
// @lc code=end

File: cpp/39.combination-sum.cpp
/*
 * @lc app=leetcode id=39 lang=cpp
 *
 * [39] Combination Sum
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> curr;
        sum(res, candidates, target, curr, 0);

        return res;
    }
    
    void sum(vector<vector<int>>& res, vector<int>& candidates, int target, vector<int>& curr, int i) {
        if(target == 0) {
            res.push_back(curr);
            return;
        }

        if(target < 0) {
            return;
        }

        for(int j = i; j < candidates.size(); j++) {
            curr.push_back(candidates[j]);
            sum(res, candidates, target - candidates[j], curr, j);
            curr.pop_back();
        }

    }

};
// @lc code=end

